Create a perpetuals swaps/trades aggregated table on Dune V2

RplusT opened the pull request:
Brief comments on the purpose of your changes:
This PR creates the initial foundation of a perpetuals swaps/trades aggregated table on Dune V2 patterned after `dex.trades`. Currently contains perpetuals protocols/markets on Optimism, namely Perpetual V2, Synthetix/Kwenta, and Pika v1, v2, v3. Structured in such a way that perpetuals markets on other chains can be added later on.

*For Dune Engine V2*
I've checked that:

* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`

soispoke commented on models/perpetuals/optimism/perpetual_v2_optimism.sql:
```diff
@@ -0,0 +1,73 @@
+	LEFT JOIN {{ source('perp_v2_optimism', 'Vault_call_getFreeCollateralByRatio') }} AS co
+		ON p.evt_tx_hash = co.call_tx_hash
+	{% if is_incremental() %}
+	WHERE p.evt_block_time >= (SELECT MAX(block_time) FROM {{ this }})
```
We've settled on having incremental filters use this convention:
	`WHERE p.evt_block_time >= date_trunc("day", now() - interval '1 week')`
	I'll point to places where this should be replaced

soispoke commented on models/perpetuals/optimism/perpetual_v2_optimism.sql:
```diff
@@ -0,0 +1,73 @@
+INNER JOIN {{ source('optimism', 'transactions') }} AS tx
+	ON perps.tx_hash = tx.hash
+	{% if not is_incremental() %}
+	AND tx.block_time >= (SELECT MIN(block_time) FROM perps)
```
Would be better to have a hardcoded date in the not is incremental condition if possible

For the incremental, need to be replaced by 
	`AND tx.block_time >= date_trunc("day", now() - interval '1 week')`

soispoke commented on models/perpetuals/optimism/pika_v1_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+		,evt_block_time
+		,evt_block_number
+		,'1' AS version
+	FROM {{ source('pika_perp_optimism', 'PikaPerpV2_evt_NewPosition') }}
```
add incremental filter here:
`{% if is_incremental() %}
WHERE evt_block_time >= date_trunc("day", now() - interval '1 week') 
{% endif %}`

soispoke commented on models/perpetuals/optimism/pika_v1_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+		,evt_block_number
+		,'1' AS version
+	FROM {{ source('pika_perp_optimism', 'PikaPerpV2_evt_ClosePosition') }}
+),
```
Same here
`{% if is_incremental() %} WHERE evt_block_time >= date_trunc("day", now() - interval '1 week') {% endif %}`

soispoke commented on models/perpetuals/optimism/pika_v1_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+		,margin * leverage AS volume_raw
+		,evt_tx_hash AS tx_hash
+		,evt_index
+	FROM positions
```
here as well:  	`{% if is_incremental() %} WHERE evt_block_time >= date_trunc("day", now() - interval '1 week') {% endif %}`

soispoke commented on models/perpetuals/optimism/pika_v1_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+INNER JOIN {{ source('optimism', 'transactions') }} AS tx
+	ON perps.tx_hash = tx.hash
+	{% if not is_incremental() %}
+	AND tx.block_time >= (SELECT MIN(block_time) FROM perps)
```
Replace by 	`WHERE tx.block_time >= date_trunc("day", now() - interval '1 week')` for incremental filter

and hardcoded value for not is_incremental one

soispoke commented on models/perpetuals/optimism/pika_v2_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+		,evt_block_time
+		,evt_block_number
+		,'2' AS version
+	FROM {{ source('pika_perp_v2_optimism', 'PikaPerpV2_evt_NewPosition') }}
```
`{% if is_incremental() %} WHERE evt_block_time >= date_trunc("day", now() - interval '1 week') {% endif %}`

soispoke commented on models/perpetuals/optimism/pika_v2_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+		,evt_block_time
+		,evt_block_number
+		,'2' AS version
+	FROM {{ source('pika_perp_v2_optimism', 'PikaPerpV2_evt_ClosePosition') }}
```
`{% if is_incremental() %} WHERE evt_block_time >= date_trunc("day", now() - interval '1 week') {% endif %}`

soispoke commented on models/perpetuals/optimism/pika_v2_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+		,margin * leverage AS volume_raw
+		,evt_tx_hash AS tx_hash
+		,evt_index
+	FROM positions
```
`{% if is_incremental() %} WHERE evt_block_time >= date_trunc("day", now() - interval '1 week') {% endif %}`

soispoke commented on models/perpetuals/optimism/pika_v2_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+INNER JOIN {{ source('optimism', 'transactions') }} AS tx
+	ON perps.tx_hash = tx.hash
+	{% if not is_incremental() %}
+	AND tx.block_time >= (SELECT MIN(block_time) FROM perps)
```
`{% if is_incremental() %} WHERE tx.block_time >= date_trunc("day", now() - interval '1 week') {% endif %}`

+ hardcoded value not is_incremental

soispoke commented on models/perpetuals/optimism/pika_v3_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+		,evt_block_time
+		,evt_block_number
+		,'3' AS version
+	FROM {{ source('pika_perp_v3_optimism', 'PikaPerpV3_evt_NewPosition') }}
```
`{% if is_incremental() %} WHERE evt_block_time >= date_trunc("day", now() - interval '1 week') {% endif %}`

soispoke commented on models/perpetuals/optimism/pika_v3_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+		,evt_block_time
+		,evt_block_number
+		,'3' AS version
+	FROM {{ source('pika_perp_v3_optimism', 'PikaPerpV3_evt_ClosePosition') }}
```
`{% if is_incremental() %} WHERE evt_block_time >= date_trunc("day", now() - interval '1 week') {% endif %}`

soispoke commented on models/perpetuals/optimism/pika_v3_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+		,margin * leverage AS volume_raw
+		,evt_tx_hash AS tx_hash
+		,evt_index
+	FROM positions
```
`{% if is_incremental() %} WHERE evt_block_time >= date_trunc("day", now() - interval '1 week') {% endif %}`

soispoke commented on models/perpetuals/optimism/pika_v3_optimism.sql:
```diff
@@ -0,0 +1,151 @@
+INNER JOIN {{ source('optimism', 'transactions') }} AS tx
+	ON perps.tx_hash = tx.hash
+	{% if not is_incremental() %}
+	AND tx.block_time >= (SELECT MIN(block_time) FROM perps)
```
`{% if is_incremental() %} WHERE tx.block_time >= date_trunc("day", now() - interval '1 week') {% endif %}`

and hardcoded value for not is_incremental

soispoke commented on models/perpetuals/optimism/synthetix_optimism.sql:
```diff
@@ -0,0 +1,115 @@
+		,s.tradeSize AS volume_raw
+		,s.evt_tx_hash AS tx_hash
+		,s.evt_index
+	FROM {{ source('synthetix_optimism', 'FuturesMarket_evt_PositionModified') }} AS s
```
use incremental filters on decoded tables here using s.evt_block_time

soispoke commented on models/perpetuals/optimism/synthetix_optimism.sql:
```diff
@@ -0,0 +1,115 @@
+INNER JOIN {{ source('optimism', 'transactions') }} AS tx
+	ON perps.tx_hash = tx.hash
+	{% if not is_incremental() %}
+	AND tx.block_time >= (SELECT MIN(block_time) FROM perps)
```
replace incremental filter here

soispoke reviewed (changes requested):
Thanks a lot for this submission @RplusT ! This is a first pass to add/modify incremental filters on all decoded/base tables across spells

@jeff-dude since @dsalv this is quite similar to dex.trades, would be great to have your opinions as well

RplusT commented:
Hi @soispoke! Thanks for reviewing and giving directions regarding the incremental filters. I've made the changes. :)

soispoke commented on models/perpetuals/optimism/pika_v2_optimism.sql:
```diff
@@ -0,0 +1,161 @@
+SELECT
+	'optimism' AS blockchain
+	,perps.block_time
+	,perps.virtual_asset
```
it looks like this column doesn't exist

RplusT replied:
> it looks like this column doesn't exist

@soispoke which column are you referring to? Or did you mean all three columns?

> If this is confusing let me know and I can commit some changes myself, happy to help out

Thanks! I got the hang of it though and made the changes. Will be ready to commit once I get the clarity on which column/s you were referring to above. :)

soispoke replied:
@RplusT Was referring specifically to the `perps.virtual_asset` column

RplusT replied:
@soispoke that's odd, it's running fine when I try out the compiled code on dune.com: https://dune.com/queries/1243365?d=1

any suggestion on where else I should look to find a fix?

soispoke replied:
@RplusT I think you're just not getting the `virtual_asset` values from your `perps` CTE ?

RplusT replied:
@soispoke the `perps` CTE derives it from a `CASE` statement involving the `productId` values from `pika_perp_v2_optimism.PikaPerpV2_evt_NewPosition` and `pika_perp_v2_optimism.PikaPerpV2_evt_ClosePosition`. 

```
,CASE
		WHEN productId = 1 OR productId = 16 THEN 'ETH'
		WHEN productId = 2 OR productId = 17 THEN 'BTC'
		WHEN productId = 3 OR productId = 18 THEN 'LINK'
		WHEN productId = 4 OR productId = 19 THEN 'SNX'
		WHEN productId = 5 OR productId = 20 THEN 'SOL'
		WHEN productId = 6 OR productId = 21 THEN 'AVAX'
		WHEN productId = 7 OR productId = 22  THEN 'MATIC'
		WHEN productId = 8 THEN 'LUNA'
		WHEN productId = 9 OR productId = 23 THEN 'AAVE'
		WHEN productId = 10 OR productId = 24 THEN 'APE'
		WHEN productId = 11 OR productId = 25 THEN 'AXS'
		WHEN productId = 12 OR productId = 26 THEN 'UNI'
		ELSE CONCAT ('product_id_', productId)
		END AS virtual_asset
```

jeff-dude replied:
started new comment below on correct file

soispoke reviewed (commented):
A general note: In order to fit a structure similar to nft.trades, can you please have schema/aliases renamed in the following way:

- perpetual_v2_optimism should be renamed perpetual_v2_optimism_trades (schema = perpetual_v2_optimism, alias = trades)
- perptual_trades (schema = perpetuals, alias=trades)
- pika_v1_optimism should be renamed pika_v1_optimism_trades (schema = pika_v1_optimism, alias =trades)
etc...

If this is confusing let me know and I can commit some changes myself, happy to help out

soispoke commented:
In the `models/perpetuals/optimism/perpetual_v2_optimism.sql`, you are calling the `perps.virtual_asset` line 46, but there is no such column in your perps CTE above (line 11 to 41), hence the error.

jeff-dude commented:
@RplusT please see this commit for changes to overall structure, naming standards, table setups, query enhancements, etc. -- i hope it all makes sense and welcome feedback!

jeff-dude commented on models/perpetual/optimism/perpetual_v2_optimism_trades.sql:
```diff
@@ -0,0 +1,82 @@
+WITH perps AS (
+	SELECT
+		p.evt_block_time AS block_time
+		,"" AS virtual_asset
```
as to @soispoke comment on missing column, i think it was an accidental comment on the wrong file. it's actually this script with the issue. notice here i applied force empty spaces for three columns, which all didn't exist but were called downstream

jeff-dude commented:
> A general note: In order to fit a structure similar to nft.trades, can you please have schema/aliases renamed in the following way:
> 
> * perpetual_v2_optimism should be renamed perpetual_v2_optimism_trades (schema = perpetual_v2_optimism, alias = trades)
> * perptual_trades (schema = perpetuals, alias=trades)
> * pika_v1_optimism should be renamed pika_v1_optimism_trades (schema = pika_v1_optimism, alias =trades)
>   etc...
> 
> If this is confusing let me know and I can commit some changes myself, happy to help out

i restructured all the files 👍

jeff-dude commented:
pushed another commit which materializes final object (perpetual.trades) as a table rather than view. this is an effort that is happening now on nft / dex trades as well #1515

jeff-dude commented:
pushed two more commits which add to general macros for display in the dune app UI explorer & optimize statements to ensure delta tables are clean

jeff-dude commented:
@RplusT while i work on building these objects in test schema (prefaced with 'dbt_jeff_'), can you look into adding a unit test for perpetual trades? for dex, we did a generic test ensuring token bought was truly the token bought and not switched with the token sold.

here is test:
https://github.com/duneanalytics/spellbook/blob/main/tests/generic/dex_trades_token_bought.sql

here is seed file to support test:
https://github.com/duneanalytics/spellbook/blob/main/seeds/dex/trades/dex_trades_seed.csv

for the seed file, you can handpick 5-10ish tx's from optimism block explorer, supply values into csv from there. sometimes a simple approach is a google sheet table exported as csv. i'm not sure token bought is a test that can be applied here, but i hope it provides a solid baseline to brainstorm a test

jeff-dude commented:
tests passed for all versions of pika & synthetix, however, perpetual_v2 failed:
```
select
    (block_time || '-' || project || '-' || version || '-' || tx_hash || '-' || evt_index) as unique_field,
    count(*) as n_records

from dbt_jeff_perpetual_v2_optimism.trades
where (block_time || '-' || project || '-' || version || '-' || tx_hash || '-' || evt_index) is not null
group by (block_time || '-' || project || '-' || version || '-' || tx_hash || '-' || evt_index)
having count(*) > 1
;


select *
from dbt_jeff_perpetual_v2_optimism.trades
where block_time = '2022-08-04 15:23:19'
and tx_hash = '0x437fac4f2ee2f4e466069d2c36c9c968c7978e2d2aa27106047c7e61d14dd0a7'
and evt_index = 9
;
```

note this is the table with forced blanks for three fields, but also the margin_usd field has different values on the "dupes"

jeff-dude commented:
all objects exist now with this naming standard:
dbt_jeff_pika_v1_optimism.trades
...
...
...
dbt_jeff_perpetual.trades

i still need to test incremental loads (the prior comment will block incremental on perpetual v2 due to dupes), but you can query them in the meantime

jeff-dude commented:
i'll be traveling to berlin tomorrow and focused on dunecon next week, but will try to help here as much as i can 🙌

RplusT commented:
@jeff-dude thank you so much for all the work especially the clean up! I've corrected the missing columns error on `perpetual_v2_optimism_trades`. Also added a unit test for `perpetual_trades` wherein it'll check the perpetuals market involved in the trade/swap. Let me know if that's okay. Thank you!

RplusT commented:
@jeff-dude regarding the reorganization, I noticed that the `perpetual_v2` files are lumped together in the same `perpetual` folder with the aggregated `perpetual` market files. Is this intentional?

I saw that `synthetix` and `pika` have their own separate protocol specific folders now. Bit confusing that a protocol is named after the market its in 😅 (perpetual protocol - perpetuals market) hence why I initially named the market in plural form ("perpetuals") to distinguish them.

jeff-dude commented:
finally back to help finalize this PR 🙌 

as for the confusion around ```perpertual``` being both protocol & sector, how about just ```perpetual_protocol``` to make it clear? not perfect, but should be obvious. i can push those changes as i test your latest commits.

jeff-dude commented:
thank you for adding a generic unit test! in order to support the test, you will need additional steps. a seed file (csv format) will need built, similar to dex: https://github.com/duneanalytics/spellbook/blob/main/seeds/dex/trades/dex_trades_seed.csv

you can handpick ~5 tx's for each protocol, but the data points in the seed file need to support all column references in the generic test you added

once the seed is built, we can then apply the test in the schema file, such as: https://github.com/duneanalytics/spellbook/blob/main/models/uniswap/ethereum/uniswap_ethereum_schema.yml#L54

jeff-dude commented:
thank you! i've pushed a commit for latest reorg, should be clean now

i'll start to rebuild objects in test env and ensure all is as expected. i'll push any necessary changes along the way

jeff-dude commented:
perpetual protocol still fails with duplicates, example below:
```
select
    (block_time || '-' || project || '-' || version || '-' || tx_hash || '-' || evt_index) as unique_field,
    count(*) as n_records

from dbt_jeff_perpetual_protocol_v2_optimism.trades
where (block_time || '-' || project || '-' || version || '-' || tx_hash || '-' || evt_index) is not null
group by (block_time || '-' || project || '-' || version || '-' || tx_hash || '-' || evt_index)
having count(*) > 1
order by 2 desc
;

select *
from dbt_jeff_perpetual_protocol_v2_optimism.trades
where (block_time || '-' || project || '-' || version || '-' || tx_hash || '-' || evt_index) = '2022-08-06 00:03:05-Perpetual-2-0x442ca237ec9342895bcd318157e83288470de7603db0491e907b7e1df5b9476e-53'
;
```

jeff-dude commented:
for awareness, i'm also looking to test the generic test/seed file. for some reason, some of them are giving me issues. for example:
```
select m.underlying_asset
    , seed.underlying_asset
    , concat(seed.underlying_asset, m.underlying_asset)
    , case
        when m.market_address = seed.market_address
            then True
        else False
    end as market_address_test
from dbt_jeff_synthetix_optimism.trades m
-- from dbt_jeff_pika_v3_optimism.trades m
inner join dbt_jeff_test_data.perpetual_trades_seed seed
    on m.tx_hash = seed.tx_hash
    and m.block_date = seed.block_date
    and m.blockchain = seed.blockchain
    -- and m.underlying_asset = seed.underlying_asset
    -- and m.market = seed.market
    and m.project = seed.project
    and m.version = seed.version
```
if i look at synthetix, when i comment out underlying_asset & market in the join, rows are returned. however, when uncommented, no rows are returned. i've verified both tables have values stored as strings & attempted to concat to see side-by-side (also tried lower/trim functions) which didn't work. as for pika v3, those seem to return fine with all columns in the join. i'm a bit stumped on these

RplusT commented:
@jeff-dude sorry about the dupes on perpetual_protocol. I addressed them with the latest commit; had to break the tie amongst multiple concurrent calls `perp_v2_optimism.Vault_call_getFreeCollateralByRatio` table

Hmm I'm also not sure why that test would fail for Synthetix. Once the fix for perpetual_protocol is okay, could we try running that test on perpetual and see if it works? Could help in isolating the problem on whether it's only with Synthetix.

jeff-dude commented:
while i look into latest change, can you add a few entries to the seed file for pika v1 (assuming its applicable)?

RplusT commented:
@jeff-dude done, added some pika v1 transactions

dune-eng commented:
Workflow run id 3091044375 approved.

jeff-dude commented:
> @jeff-dude done, added some pika v1 transactions

thank you. i also edited the seed file to be 'block_date' vs. 'block_time' -- the reason being, the timestamp value you see in the dune web app isn't how it's actually stored in the database. it's a cleaner view on the app, but you'd need a more specific format for the join. date makes it much easier for us.

jeff-dude commented:
> @jeff-dude sorry about the dupes on perpetual_protocol. I addressed them with the latest commit; had to break the tie amongst multiple concurrent calls on `perp_v2_optimism.Vault_call_getFreeCollateralByRatio` table
> 
> Hmm I'm also not sure why that test would fail for Synthetix. Once the fix for perpetual_protocol is okay, could we try running that test on perpetual and see if it works? Could help in isolating the problem on whether it's only with Synthetix.

the duplicate issue appears to be fixed. i'll look to do an incremental load test too. the outstanding issue is around that generic test on synthetix. it's an odd issue. the test "passes", but that is only because no rows are returned in the join from table to seed

jeff-dude commented:
@RplusT i believe the issue around generic test on synthetix comes down to the two columns in the join ```market``` and ```underlying_asset``` containing extra characters that aren't seen in the UI.

in the model to build synthetix table, you have logic for each of these columns to convert from hex to binary, then decode on top. if you run a length() function on these columns, you get 32. in the seed file, since values are hardcoded, you get the true string length of 3, 4 or 5. these hidden characters break the join.
```
select
  market,
  length(market) 
  -- , hex(market)
  -- , concat('0x', hex(market))
  -- , encode(concat('0x', hex(market)), 'US-ASCII')
from
  dbt_jeff_synthetix_optimism.trades
```

do you know of the spark sql function which would basically do the opposite of logic in your model & help us convert back to string for the join to hardcoded seed file? or a way to strip off end characters?

RplusT commented:
> do you know of the spark sql function which would basically do the opposite of logic in your model & help us convert back to string for the join to hardcoded seed file? or a way to strip off end characters?

@jeff-dude I'm a bit stumped here. Been trying to use `BTRIM`, `TRIM` and the other variations to remove the trailing spaces, along with `SUBSTRING` but no dice. Also tried `REGEXP_EXTRACT` but it doesn't seem to work on Dune? Will keep digging.

Is changing the test also an option? Maybe with columns that aren't problematic hehe

jeff-dude commented:
> > do you know of the spark sql function which would basically do the opposite of logic in your model & help us convert back to string for the join to hardcoded seed file? or a way to strip off end characters?
> 
> @jeff-dude I'm a bit stumped here. Been trying to use `BTRIM`, `TRIM` and the other variations to remove the trailing spaces, along with `SUBSTRING` but no dice. Also tried `REGEXP_EXTRACT` but it doesn't seem to work on Dune? Will keep digging.
> 
> Is changing the test also an option? Maybe with other columns? Although I worry about the implications later on when other wizards try doing a join on these columns.

i'll try to ask some teammates as well. hopefully we can find an approach.

if the generic test still works after removing those two columns from the join, and you don't foresee downstream use cases of this spell being joined to other spells/base tables/etc on these two columns, we can likely go that route

jeff-dude commented:
@RplusT this query shows how we can tweak the join conditions to make rows returned as expected:
```
select
  m.market,
  length(m.market),
  hex(cast(m.market as binary)),
  replace(hex(cast(m.market as binary)), '0', ''),
  seed.market,
  length(seed.market),
  hex(cast(seed.market as binary))
from
  dbt_jeff_synthetix_optimism.trades m
  inner join dbt_jeff_test_data.perpetual_trades_seed seed on m.tx_hash = seed.tx_hash
  and m.block_date = seed.block_date
  and m.blockchain = seed.blockchain 
  and replace(hex(cast(m.underlying_asset as binary)), '0', '') = hex(cast(seed.underlying_asset as binary))
  and replace(hex(cast(m.market as binary)), '0', '') = hex(cast(seed.market as binary))
  and m.project = seed.project
  and m.version = seed.version
```

my open question is: will this affect other protocols outside of synthetix that don't have this issue? i should be able to test that with objects in this PR. 

the other consideration is your downstream usage of these columns in queries, dashboards, etc. -- if these two columns are used, you will need to remember to apply this type of function usage to make it work

jeff-dude reviewed (approved):
okay @RplusT i've finalized this to where i believe this is in a good spot 👍 i removed those two columns from the join condition -- as the unique key on each model doesn't include them, shouldn't be needed, and tests proved to be fine. just keep in mind the comment where if you use the columns downstream, they may be impacted on joins/filters.

please also note other recent commits to see updates on enhancing spells, things we've learned in the last week or so.

thanks again for all the back & forth, this should have the opportunity to become an important spell that spans many projects & blockchains!

RplusT commented:
@jeff-dude thank you so much for finalizing this and all your work beforehand on it (@soispoke too)! Really appreciate the guidance.

I'm looking forward to building some dashboards on top of it. Hopefully other wizards are too. :)

> please also note other recent commits to see updates on enhancing spells, things we've learned in the last week or so.

got it, will do. :)

RplusT commented:
Hi @jeff-dude! Not sure where it was best to reach you so my apologies for doing it here. I tried to query the perpetuals.trades table on dune.com but it's saying the table `perpetuals.trades` does not exist. I also tried finding it via the menu on the left side but the Spells > Optimism > `perpetual` sector > `trades` table is showing up as `perpetual_protocol_v2_optimism.trades`. Would you know the possible reason/s behind this?

jeff-dude commented:
> Hi @jeff-dude! Not sure where it was best to reach you so my apologies for doing it here. I tried to query the perpetuals.trades table on dune.com but it's saying the table `perpetuals.trades` does not exist. I also tried finding it via the menu on the left side but the Spells > Optimism > `perpetual` sector > `trades` table is showing up as `perpetual_protocol_v2_optimism.trades`. Would you know the possible reason/s behind this?

try removing the 's' in schema

```
select count(1)
from perpetual.trades
;
```

but you have a good point on the explorer displaying the incorrect objects. we'll have to look closer at that. it has to do with the post-hook property in the config block, which calls the ```expose_spells``` macro. that macro tells spark how to display. we're missing something in how we provide info on that

RplusT commented:
> try removing the 's' in schema

@jeff-dude ah got it! That works. Realized there was a bug too while exploring so did the fix on a separate PR 😅

> but you have a good point on the explorer displaying the incorrect objects...

Understood. Anything I can do to help out there or is that an internal thing you'll look at on your end?
